File: leetcode/New21Game/a.cpp
/* Copyright (C) 2018 by iamslash */

#include <cstdio>
#include <vector>
#include <algorithm>
#include <numeric>

// 4ms 82.61% 9.5MB 58.33%
// prefix sum
// O(N) O(N)
class Solution {
 public:
  double new21Game(int N, int K, int W) {
    if (K == 0 || N >= K+W)
      return 1.0;
    std::vector<double> P(N+1, 0.0);
    P[0] = 1.0;
    double sum = 1.0;
    for (int i = 1; i <= N; ++i) {
      P[i] += sum / W;
      if (i < K)
        sum += P[i];
      if (i - W >= 0)
        sum -= P[i-W];
    }
    return std::accumulate(P.begin() + K,
                           P.end(),
                           0.0);        
  }
};

int main() {
  return 0;
}

File: leetcode/New21Game/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2020 by iamslash 

# 72ms 79,53% 14.5MB 21.14%
# iterative dynamic programming
# O(N) O(N)
class Solution:
    def new21Game(self, N: int, K: int, W: int) -> float:
        if K == 0 or N >= K + W:
            return 1.0
        C = [1.0] + [0.0] * N
        psum = 1.0
        for i in range(1, N+1):
            C[i] = psum / W
            if i < K:
                psum += C[i]
            if i - W >= 0:
                psum -= C[i-W]
        return sum(C[K:])
                
if __name__ == "__main__":
    sln = Solution()
    print(sln.new21Game(10, 1, 10)) # 1.00000
    print(sln.new21Game(6, 1, 10)) # 0.60000
    print(sln.new21Game(21, 17, 10)) # 0.73278
